ocaml/xapi/xapi_host.ml

```diff
@@ -2904,14 +2905,97 @@ let apply_updates ~__context ~self ~hash =
+    (fun g ->
+      [
+        Api_errors.updates_require_recommended_guidance
+      ; Updateinfo.Guidance.to_string g
+      ]
+    )
+    guidances
+  @ warnings
 
 let set_https_only ~__context ~self ~value =
   let state = match value with true -> "close" | false -> "open" in
   ignore
   @@ Helpers.call_script !Xapi_globs.firewall_port_config_script [state; "80"] ;
   Db.Host.set_https_only ~__context ~self ~value
+
+let restart_device_models_for_recommended_guidances ~__context ~host =
+  (* Restart device models of all running HVM VMs on the host by doing
+   * local migrations if it is required by recommended guidances. *)
+  Repository_helpers.do_with_device_models ~__context ~host
+    ~only_vm_with_recommended_guidances:true
+  @@ fun (ref, record) ->
+  match
+    (record.API.vM_power_state, Helpers.has_qemu_currently ~__context ~self:ref)
+  with
+  | `Running, true ->
+      Xapi_vm_migrate.pool_migrate ~__context ~vm:ref ~host
+        ~options:[("live", "true")] ;
+      None
+  | `Paused, true ->
+      error "VM 'ref=%s' is paused, can't restart device models for it"
+        (Ref.string_of ref) ;
+      Some ref
+  | _ ->
+      (* No device models are running for this VM *)
+      None
+
+let apply_recommended_guidances ~__context ~self =
+  try
+    let open Repository_helpers in
+    let open Updateinfo in
+    let open Updateinfo.Guidance in
+    let host_update_guidances =
+      Db.Host.get_recommended_guidances ~__context ~self
+    in
+    let host_guidances =
+      List.map
+        (fun ug ->
+          match ug with
+          | `reboot_host ->
+              Guidance.RebootHost
+          | `reboot_host_on_livepatch_failure ->
+              Guidance.RebootHostOnLivePatchFailure
+          | `restart_toolstack ->
+              Guidance.RestartToolstack
+          | `restart_device_model ->
+              Guidance.RestartDeviceModel
+        )
+        host_update_guidances
+    in
+    match host_guidances with
+    | [] ->
+        (* try to restart device models for vm recommended_guidances*)
+        restart_device_models_for_recommended_guidances ~__context ~host:self
+    | [RebootHost] ->
+        reboot ~__context ~host:self
+    | [EvacuateHost] ->
+        (* EvacuatHost should have been done before applying updates by XAPI users.
+         * Here only the guidances to be applied after applying updates are handled,
+         * nothing need to be done for EvacuatHost.
+         *)
+        ()
+    | [RestartToolstack] ->
+        (* try to restart device models for vm recommended_guidances*)
+        restart_device_models_for_recommended_guidances ~__context ~host:self ;
+        restart_agent ~__context ~host:self
+    | l when GuidanceSet.eq_set1 l ->
+        (* EvacuateHost and RestartToolstack *)
+        restart_agent ~__context ~host:self
+    | l ->
+        let host' = Ref.string_of self in
+        error
+          "Found wrong guidance(s) after applying updates on host ref='%s': %s"
```

Comment:
The context `after applying updates` would not be applicable since it is in a separated API.

Reply:
error log updated